lib/awful/keygrabber.lua

```diff
@@ -1,22 +1,96 @@
 ---------------------------------------------------------------------------
---- Keygrabber Stack
+--- Extra helper functions for the keygrabber.
 --
 -- @author dodo
+-- @author Emmanuel Lepage Vallee &lt;elv1313@gmail.com&gt;
 -- @copyright 2012 dodo
--- @module awful.keygrabber
+-- @copyright 2017 Emmanuel Lepage Vallee
+-- @classmod keygrabber
 ---------------------------------------------------------------------------
 
 local ipairs = ipairs
 local table = table
-local capi = {
-    keygrabber = keygrabber }
+local gdebug = require('gears.debug')
+local akey = require("awful.key")
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+local gtable = require("gears.table")
+local gobject = require("gears.object")
+local gtimer = require("gears.timer")
+local capi = { keygrabber = keygrabber, root = root }
 
-local keygrabber = {}
+local keygrab = {}
 
 -- Private data
 local grabbers = {}
 local keygrabbing = false
 
+local keygrabber = {
+    object = {}
+}
+
+-- Instead of checking for every modifiers, check the key directly.
+local conversion = {
+    Super_L   = "Mod4",
+    Control_L = "Control",
+    Shift_L   = "Shift",
+    Alt_L     = "Mod1",
+    Super_R   = "Mod4",
+    Control_R = "Control",
+    Shift_R   = "Shift",
+    Alt_R     = "Mod1",
+}
+
+--BEGIN one day create a proper API to add and remove keybindings at runtime.
+-- Doing it this way is horrible.
+
+-- Delay until the next loop otherwise `rc.lua: root.keys(globalkeys)` will
+-- overwrite them if the keygrabber is created directly from `rc.lua`.
+local delay_list = {}
+
+local function add_root_keybindings(self, list)
+    if not list then return end
+
+    local was_started = #delay_list > 0
+
+    -- Changing all keybindings is very slow and freezes Awesome for a visible
+    -- amount of timer. Make sure to batch as much as possible.
```

Comment:
Benchmark:
```diff
diff --git a/tests/test-benchmark.lua b/tests/test-benchmark.lua
index def3c01f..a6f5cbff 100644
--- a/tests/test-benchmark.lua
+++ b/tests/test-benchmark.lua
@@ -71,11 +71,30 @@ local function e2e_tag_switch()
     do_pending_repaint()
 end
 
-benchmark(create_and_draw_wibox, "create&draw wibox")
-benchmark(update_textclock, "update textclock")
-benchmark(relayout_textclock, "relayout textclock")
-benchmark(redraw_textclock, "redraw textclock")
+local keys
+local function setup_keys(n)
+    keys = {}
+    local gears = require("gears")
+    for i=1, n do
+        keys = gears.table.join(keys, awful.key({"mod4"}, "x", function() end, { description = "foo", group="bar" }))
+    end
+end
+
+local function change_keybindings()
+    root.keys(keys)
+end
+
+--benchmark(create_and_draw_wibox, "create&draw wibox")
+--benchmark(update_textclock, "update textclock")
+--benchmark(relayout_textclock, "relayout textclock")
+--benchmark(redraw_textclock, "redraw textclock")
 benchmark(e2e_tag_switch, "tag switch")
+setup_keys(10)
+benchmark(change_keybindings, "change keys 10")
+setup_keys(100)
+benchmark(change_keybindings, "change keys 100")
+setup_keys(1000)
+benchmark(change_keybindings, "change keys 1000")
 
 runner.run_steps({ function() return true end })
 
```
Result:
```
          tag switch: 0.00571348 sec/iter (180 iters, 1.354 sec for benchmark)
      change keys 10: 0.00265768 sec/iter (380 iters, 1.013 sec for benchmark)
     change keys 100: 0.0261727  sec/iter ( 39 iters, 2.046 sec for benchmark)
    change keys 1000: 0.261728   sec/iter (  4 iters, 1.31 sec for benchmark)
```
That's less than 0.3 milliseconds per key and this seems to scale linearly with the number of keys. Sorry, but what are you doing that this turns into visible freezes?!?